server/base/src/main/java/org/apache/accumulo/server/util/UpgradeUtil.java

```diff
@@ -57,79 +76,222 @@ public String keyword() {
+        // modified to account for the new Fate table.
+        if (!zr.getChildren(Constants.ZFATE).isEmpty()) {
+          throw new IllegalStateException("Cannot continue pre-upgrade checks"
+              + " because FATE transactions exist. You can start a tserver, but"
+              + " not the Manager, with the old version of Accumulo then use "
+              + " the shell to delete completed transactions and fail pending"
+              + " or in-progress transactions. Once all of the FATE transactions"
+              + " have been removed you can retry this operation.");
+        }
+      } catch (KeeperException | InterruptedException e) {
+        throw new IllegalStateException("Error checking for existing FATE transactions", e);
+      }
+      LOG.info("No FATE transactions found");
+
+      // Forcefully delete all server locks
+      Set<ServiceLockPath> serviceLockPaths =
+          context.getServerPaths().getCompactor((g) -> true, AddressSelector.all(), true);
+      serviceLockPaths.addAll(
+          context.getServerPaths().getTabletServer((g) -> true, AddressSelector.all(), true));
+      serviceLockPaths
+          .addAll(context.getServerPaths().getScanServer((g) -> true, AddressSelector.all(), true));
+      var mgrPath = context.getServerPaths().getManager(true);
+      if (mgrPath != null) {
+        serviceLockPaths.add(mgrPath);
+      }
+      var gcPath = context.getServerPaths().getGarbageCollector(true);
+      if (gcPath != null) {
+        serviceLockPaths.add(gcPath);
+      }
+      var monitorPath = context.getServerPaths().getMonitor(true);
+      if (monitorPath != null) {
+        serviceLockPaths.add(monitorPath);
+      }
+
+      for (ServiceLockPath slp : serviceLockPaths) {
+        LOG.info("Deleting all zookeeper entries under {}", slp);
+        try {
+          List<String> children = zr.getChildren(slp.toString());
+          for (String child : children) {
+            LOG.debug("Performing recursive delete on node:  {}", child);
+            ZooUtil.recursiveDelete(zs, slp + "/" + child, NodeMissingPolicy.SKIP);
+          }
+        } catch (KeeperException.NoNodeException e) {
+          LOG.warn("{} path does not exist in zookeeper", slp);
+        } catch (KeeperException e) {
+          throw new IllegalStateException(
+              "Error performing recursive delete on children under node: " + slp.toString(), e);
+        } catch (InterruptedException e) {
+          throw new IllegalStateException("Interrupted while trying to find"
+              + " and delete children of zookeeper node: " + slp);
+        }
+      }
+    }
+
+    // Initialize the UpgradeProgress object in ZooKeeper. If the node exists, maybe
+    // because the 'check' command is being re-run, delete it.
+    try {
+      if (zr.exists(Constants.ZUPGRADE_PROGRESS)) {
+        ZooUtil.recursiveDelete(zs, Constants.ZUPGRADE_PROGRESS, NodeMissingPolicy.FAIL);
```

Comment:
Could use the force option here.

```suggestion
        if(force){
           ZooUtil.recursiveDelete(zs, Constants.ZUPGRADE_PROGRESS, NodeMissingPolicy.FAIL);
         } else {
            throw new IllegalstateException("An upgrade was initiated and may be in progress, use -force option to delete upgrade progress tracking");
         }
```

Reply:
In https://github.com/apache/accumulo/pull/5438#discussion_r2029068401, @ctubbsii suggested that the `--prepare` command should be idempotent. I was trying to take the same approach by just removing the node if for some reason the user ran the `--check` command again.

Reply:
I was thinking of the case where the check command was run, the manager started, then the check command run again.  In that case deleting it may be dicey.  The second run of the check command could be completely accidental run from a shell history miss.

Reply:
If the Manager is started and an upgrade is required, then the call to UpgradeProgressTracker.continueUpgrade will throw an exception if the call to UpgradeProgressTracker.initialize (below) has not been called. I think in this case the Manager will fail to start.

Reply:
Yeah, thinking of a case like the following

 1. User runs --check and it succeeds.
 2. User starts manager and other serer processes
 3. Manager starts working on upgrade because the progress tracking stuff exists in zk
 4. user accidentally runs the --check command again and it delete the upgrade progress tracking stuff in ZK

Reply:
I added logic in adca5cd to prevent the `check` command from being run if the Manager lock is held.